src/vscode-bicep/src/commands/decompile.ts

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import {
+  IActionContext,
+  UserCancelledError,
+} from "@microsoft/vscode-azext-utils";
+import assert from "assert";
+import * as fse from "fs-extra";
+import vscode, { MessageItem, Uri, window } from "vscode";
+import { DocumentUri, LanguageClient } from "vscode-languageclient/node";
+import { OutputChannelManager } from "../utils/OutputChannelManager";
+import { Command } from "./types";
+import path from "path";
+
+interface DecompileCommandParams {
+  jsonUri: DocumentUri;
+}
+
+interface DecompiledFile {
+  // absolute path to overwrite - could be outside input file's folder/subfolders (e.g. ".." in module path)
+  absolutePath: DocumentUri;
+  // relative path if choosing to copy to new folder - original paths outside input file's folder/subfolders may be renamed/moved
+  clonableRelativePath: DocumentUri;
+  bicepContents: string;
+}
+
+interface BicepDecompileCommandResult {
+  decompileId: string; // Used to synchronize telemetry events
+  output: string;
+  errorMessage?: string;
+  outputFiles: DecompiledFile[]; // First is assumed to be main output file
+  conflictingOutputPaths: DocumentUri[]; // client should verify overwrite with user
+}
+
+interface BicepDecompileSaveCommandParams {
+  decompileId: string;
+  outputFiles: DecompiledFile[];
+  overwrite: boolean;
+}
+
+interface BicepDecompileSaveCommandResult {
+  output: string;
+  errorMessage?: string;
+  mainSavedBicepPath?: string;
+  savedPaths: string[];
+}
+
+export class DecompileCommand implements Command {
+  public readonly id = "bicep.decompile";
+  public constructor(
+    private readonly client: LanguageClient,
+    private readonly outputChannelManager: OutputChannelManager
+  ) {
+    // nothing to do
+  }
+
+  public async execute(
+    context: IActionContext,
+    documentUri?: vscode.Uri | undefined
+  ): Promise<void> {
+    documentUri = documentUri ?? window.activeTextEditor?.document.uri;
+    if (!documentUri) {
+      throw new Error(
+        "Please open a JSON ARM Template file before running this command"
+      );
+    }
+
+    const canDecompile = await DecompileCommand.mightBeArmTemplate(documentUri);
+    if (!canDecompile) {
+      throw new Error(
```

Comment:
Should we just write to the Bicep Operations channel?

Reply:
[sw/decompile2 aaad77fbf] CR fix

Tough to decide.  A throw will end up as failure in the telemetry, just returning will end up as succeeded.  Ended up logging but then throwing cancel (will be cancel in telemetry).

Reply:
It's also a bit confusing as to whether we allow the language server or the extension to display error messages.  Might be something we need to make more consistent later.  Right now I'm assuming language server will do it because in some cases (e.g. Insert Resource extension isn't really involved and lang server is currently displaying the errors).